### PR TITLE
Add tidy check to error on new `Makefile`s in `tests/run-make`

### PR DESCRIPTION
**Pull request by jieyouxu**
We would like to strongly encourage new `run-make` tests to be written in Rust and not add any new `Makefile`-based `run-make` tests.

### CODE REVIEW

File: src/tools/tidy/src/run_make_tests.rs [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+use std::fs::File;
+use std::io::Write;
+use std::path::{Path, PathBuf};
+
+pub fn check(root_path: &Path, bless: bool, bad: &mut bool) {
+    let tests_path = &root_path.join("tests");
+
+    let allowed_makefiles = {
+        let allowed_makefiles = include!("expected_run_make_makefiles.txt");
+        let is_sorted = allowed_makefiles.windows(2).all(|w| w[0] < w[1]);
```
Comment (Nilstrieb) [inline src/tools/tidy/src/run_make_tests.rs:13]: i think you messed up something here, you arent actually parsing the file
can you just reuse the code from the issues test stuff?

Comment (Nilstrieb) [reply src/tools/tidy/src/run_make_tests.rs:13]: oh, `include!` does rust syntax, please leave a comment about that lol

File: src/tools/tidy/src/main.rs [hunk 2]
```diff
@@ -103,6 +103,7 @@ fn main() {
         check!(tests_revision_unpaired_stdout_stderr, &tests_path);
         check!(debug_artifacts, &tests_path);
         check!(ui_tests, &root_path, bless);
+        check!(run_make_tests, &root_path, bless);
```
[the remark this answers is not in the document]
Comment (jieyouxu) [reply src/tools/tidy/src/main.rs:106]: I'll make that change.

[the remark this answers is not in the document]
Comment (jieyouxu) [reply src/tools/tidy/src/main.rs]: Passed them separately.

File: src/tools/tidy/src/run_make_tests.rs (hunk 1 above)
Comment (jieyouxu) [reply src/tools/tidy/src/run_make_tests.rs:14]: Added a comment explaining why `include!` is used.

**Review by onur-ozkan — COMMENTED** [review]
How often will this check be useful (finding newly created Makefiles)? I think it's so rare. How about creating a shell script (which can be used in `mingw-check-tidy` pipeline) that handles this? Making additions on tidy for this check seems like an overkill to me.

**Comment by jieyouxu** [comment]
> How often will this check be useful (finding newly created Makefiles)? I think it's so rare.

There's already 3 instances where people weren't aware of the new infrastructure and tried to introduce new Makefiles, e.g. #122757, #122663, #122840, so it does happen.

> How about creating a shell script (which can be used in mingw-check-tidy pipeline)

That sounds reasonable, though I'm not very keen on writing shell scripts :3 If this seems unnecessary, it's fine if I just occasionally remind people who introduce new Makefiles to use `rmake.rs` instead (after all, the plan is to get rid of all the Makefiles under `run-make`), and I'd imagine it's going to happen less as more tests are has `rmake.rs` instead of Makefiles.

**Comment by onur-ozkan** [comment]
> I'd imagine it's going to happen less as more run-make tests become `rmake.rs` instead of Makefiles.

I thought the same and that's why I suggest creating shell script instead of bloating tidy.

**Comment by jieyouxu** [comment]
> I thought the same and that's why I suggest creating shell script instead of bloating tidy.

Now that I think about it, it's probably not even worth the shell script, because we're planning to migrate away from Makefiles anyway, and if I add a shell script it invariably is going to require some allowlist for the Makefiles, which then means anyone who's porting the run-make tests have additional friction needing to adjust that. Given that, I'm going to close this PR, thanks for the advice!

**Comment by jieyouxu** [comment]
I'm actually going to re-open this because a new Makefile slipped through in #118644.

> How about creating a shell script (which can be used in mingw-check-tidy pipeline) that handles this?

The logic for handling entries in the allowlist for run-make tests is not exactly trivial to write or maintain in a shellscript (this includes checking for sorted entries, moved/removed entries, walking the run-make tests directory).

Furthermore, I would think that it's nicer for people working on PRs to see the "don't add a new Makefile" warning locally so it can be caught early, without having to rely on PR CI to then immediately fail anyway.

**Comment by jieyouxu** [comment]
(Failing as intended, it's the new Makefile that slipped through)

**Review by albertlarsan68 — COMMENTED** [review]
A few comments

File: src/tools/tidy/src/allowed_run_make_makefiles.txt [hunk 3]
```diff
@@ -0,0 +1,344 @@
+/*
```
[the remark this answers is not in the document]
Comment (jieyouxu) [reply src/tools/tidy/src/allowed_run_make_makefiles.txt:1]: I could just change this to a simple line-separated plain text file and not rust. FWIW, this is also what the ui test tidy check does.

[the remark this answers is not in the document]
Comment (jieyouxu) [reply src/tools/tidy/src/allowed_run_make_makefiles.txt]: Changed the allowlist txt file to be actually a plain-text file.

File: src/tools/tidy/src/main.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (jieyouxu) [reply src/tools/tidy/src/main.rs]: (This was previously changed to `check!(run_make_tests, &tests_path, &src_path, bless);`)

**Comment by onur-ozkan** [comment]
I agree that this check is useful, but I'm not sure if it's worth adding this complexity on tidy. I still think we can handle this using shell scripting script to perform the check in CI.

> That sounds reasonable, though I'm not very keen on writing shell scripts :3

I can make a P.o.C PR in couple days if you don't want to do that.

**Comment by jieyouxu** [comment]
> I can make a P.o.C PR in couple days if you don't want to do that.

(That'd be great, it's moreso that I don't know how to write this in shell script lol)

**Comment by albertlarsan68** [comment]
I want to point out that we have a check in tidy that the tests aren't added to src/test, even though it isn't really needed anymore.

I prefer the rust version over a shell version, just because it is more portable, and that we have most of the infrastructure needed to run this kind of test.
It is in the name, tidy is used to keep the repo well... Tidy.

File: src/tools/tidy/src/run_make_tests.rs [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+    } else {
+        for file in remaining_makefiles {
+            let mut p = PathBuf::from(tests_path);
+            p.push(file);
+            tidy_error!(
+                bad,
+                "Makefile `{}` no longer exists and should be removed from the exclusions in \
+                `src/tools/tidy/src/allowed_run_make_makefiles.txt`",
+                p.display()
+            );
```
Comment (albertlarsan68) [inline src/tools/tidy/src/run_make_tests.rs:85]: Maybe also point to --bless here?

**Comment by workingjubilee** [comment]
We can also justifiably remove the tidy check when we drive out our foe (Makefiles).

File: src/tools/tidy/src/run_make_tests.rs [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+            let makefile_path = makefile_path.to_str().unwrap().replace('\\', "/");
+
+            if !remaining_makefiles.remove(&makefile_path) {
+                tidy_error!(
+                    bad,
+                    "found run-make Makefile not permitted in \
+                `src/tools/tidy/src/allowed_run_make_makefiles.txt`, please write new run-make \
+                tests with `rmake.rs` instead: {}",
+                    entry.path().display()
+                );
```
Comment (albertlarsan68) [inline src/tools/tidy/src/run_make_tests.rs:58]: Am I right that this error is not fixed by running with --bless?

**Comment by onur-ozkan** [comment]
> > I can make a P.o.C PR in couple days if you don't want to do that.
> 
> (That'd be great, it's moreso that I don't know how to write this in shell script lol)
> 
> Although, if it's not written in tidy, would people porting run-make tests need to manually edit some allowlist file (i.e. no `./x test tidy --bless`?)? Or is it going to look more like `./x run src/ci/some-shell-script.sh --bless`?

Manually editing. It should be easy to remove a couple of lines (we will check if those lines exist on the filesystem anyway).

> I want to point out that we have a check in tidy that the tests aren't added to src/test, even though it isn't really needed anymore.
> 
> I prefer the rust version over a shell version, just because it is more portable, and that we have most of the infrastructure needed to run this kind of test. It is in the name, tidy is used to keep the repo well... Tidy.

`tidy` is probably the most invoked command by everyone. To keep it light and quick, it shouldn't be bloated unless absolutely necessary. Maybe I am missing something about the check we are discussing; as I don't really see it as necessary to add to tidy.

File: src/tools/tidy/src/run_make_tests.rs (hunk 5 above)
Comment (jieyouxu) [reply src/tools/tidy/src/run_make_tests.rs:58]: Yes, that is the intention: that new Makefiles not in the allowlist will error, even if you pass --bless.

**Comment by albertlarsan68** [comment]
> `tidy` is probably the most invoked command by everyone. To keep it light and quick, it shouldn't be bloated unless absolutely necessary. Maybe I am missing something about the check we are discussing; as I don't really see it as necessary to add to tidy.

The check will become lighter over time, and we can remove it once the run-make with Makefiles infrastructure is removed.
Do you mind leaving a FIXME somewhere in the run-make code @jieyouxu ?

**Comment by workingjubilee** [comment]
@onur-ozkan I am not sure I see the point between tidy and another shell script here.

Anything that people should be running every time they commit or push a PR is gonna be in that hot path, and this won't catch anything if it doesn't run in that hot path, script or no, but it will require many more hours for the back-and-forth of "this should be fixed" "oh" re: a new Makefile appearing in a PR.

File: src/tools/tidy/src/run_make_tests.rs (hunk 4 above)
Comment (jieyouxu) [reply src/tools/tidy/src/run_make_tests.rs]: Updated the message to point to `--bless`

**Comment by jieyouxu** [comment]
> Do you mind leaving a FIXME somewhere in the run-make code

Left a FIXME to remind removing this check once Makefiles are no longer accepted.

**Comment by onur-ozkan** [comment]
> @onur-ozkan I am not sure I see the point between tidy and another shell script here.
> 
> Anything that people should be running every time they commit or push a PR is gonna be in that hot path, and this won't catch anything if it doesn't run in that hot path, script or no, but it will require many more hours for the back-and-forth of "this should be fixed" "oh" re: a new Makefile appearing in a PR.

The catch is rare (only when writing tests in a Makefile). A simple script can easily perform this check and can be removed right away without changing the codebase once we are done with the Makefiles. I don't want to talk about this indefinitely; it could be just my personal preference. Since most people here prefer the current approach, you can ignore my comments.

**Comment by jieyouxu** [comment]
> The catch is rare (only when writing tests in a Makefile). A simple script can easily perform this check and can be removed right away without changing the codebase once we are done with the Makefiles. I don't want to talk about this indefinitely; it could be just my personal preference. Since most people here prefer the current approach, you can ignore my comments.

I have no strong opinions on this, as I can understand that if tidy is slow it will be very annoying. I personally don't know how to write a shell script version of this, but if someone writes a shell script version of this instead that'll work too.

**Comment by albertlarsan68** [comment]
As we will have to remove the makefile-handling code, we can at the same time remove the tidy part.

r=me when CI green

**Review by albertlarsan68 — APPROVED** [review]

**Comment by jieyouxu** [comment]
> r=me when CI green

(You will need to r+ since I don't have r+)

**Comment by albertlarsan68** [comment]
Thanks for the PR!
@bors r+ p=1 (this PR may fail because of the checks it adds, but enforces a new policy)

The r=me is also a signal to other r+ people that they can approve the PR in my name, not only to the author of the PR.

**Pull request merged**